### ISSUE TITLE
"Unable to init server" warning with GTK import

### ISSUE THREAD

**Issue by mgrabovsky** [PyCQA/pylint#5251]
### Bug description

```python
When running Pylint on code which imports any GTK or GDK stuff via the GObject mechanism, the warning `Unable to init server: Could not connect: Connection refused` is printed out.

Here's a complete reproducer (running in a fresh Fedora 35 container):

    # dnf install --refresh -y python3-pylint python3-gobject pygobject3-devel gtk3-devel
    # cd /tmp ; cat > test.py <<EOF
    import sys
    print(len(sys.argv))
    EOF

`test.py` is a simple Python script with no complicated logic. Pylint goes through it just fine (with the expected :

    # pylint-3 -d C0114,W0611 test.py
    
    ------------------------------------
    Your code has been rated at 10.00/10

Now let's add the GTK import and see what happens:

    # sed -ie '1a import gi.repository.Gtk' test.py
    # pylint-3 -d C0114,W0611 test.py
    Unable to init server: Could not connect: Connection refused
    Unable to init server: Could not connect: Connection refused
    
    --------------------------------------------------------------------
    Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)

Via some magical mechanism, this also causes our Pylint pipeline to time out, see, for example, [this recent run](https://github.com/abrt/gnome-abrt/runs/4090912194?check_suite_focus=true).

I've found a related bug, #2649, but its outcome is not satisfactory.
```

### Command used

```shell
See above.
```

### Pylint output

```shell
See above.
```

### Expected behavior

I would expect Pylint to lint the file without emitting any unrelated warnings and not to time out when being run as part of a GitHub workflow.

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.4
Python 3.10.0 (default, Oct  5 2021, 00:00:00) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)]
```

### OS / Environment

Fedora 35 (inside a container)

**Comment by nozzy123nozzy** [comment]
@mgrabovsky   I'm not the maintainer of "pylint," but I confirmed your problem in my environment.  
  Your problem causes the gdk to check the alive of the display server using gdk_display_open() called from Gtk.init_check() when "test.py" imports "gi.repository.Gtk". In order to obtain all exported function lists through GObject introspection, the "astroid" internally used in "pylint," try real import of "gi.repository.Gtk" in _import_gi_module() in astroid/brain/brain_gi.py. If no display server exists in the test environment, gdk_display_open() complains, "Unable init server..."
  IMHO, the ideal solution to your problem is that the "astroid" does not real import GObject introspection, but it appears to need much work to modify. If you want to solve this problem right away, I suggest a workaround as below,
```
  env GDK_BACKEND=- pylint-3 -d C0114,W0611 test.py
```
This workaround can skip checking the display server in gdk_display_open() of gdk, which disappears the error message you mentioned.
I hope this is your help.

**Comment by nozzy123nozzy** [comment]
@mgrabovsky  I examined a bit about the reason why the Github workflow timed out. For now, I'll write something that I found.
The astroid, which uses internally in pylint, parse all the exported function definition of the imported module. Then I examined how many functions astroid has to parse about "gi.repository.Gtk." Surprisingly its result is  49,808 functions!  I'll show it as below,
```
$ env GDK_BACKEND=- python3 -mpdb (command -v pylint) -d C0114,W0611 test.py
(Pdb) b /home/mine/.local/lib/python3.9/site-packages/astroid/brain/brain_gi.py:214                                   
Breakpoint 1 at /home/mine/.local/lib/python3.9/site-packages/astroid/brain/brain_gi.py:214                           
(Pdb) c                                                                                                                 
> /home/mine/.local/lib/python3.9/site-packages/astroid/brain/brain_gi.py(214)_import_gi_module()                     
-> astng = AstroidBuilder(AstroidManager()).string_build(modcode, modname)                                              
(Pdb)  len([l for l in modcode.splitlines() if "def " in l])                                                            
49808                              
```
 The gnome-abrt, which you try to examine with pylint, includes several modules using gi. It's just my guess, and this import causes time out to parse a massive number of functions, 
 At this moment, I can't come up with any workarounds about this problem.

**Comment by Pierre-Sassoulas** [comment]
Using the ``limit-inference-results`` option might help in this case, could you try and see if it solves the problem, please ?

**Comment by nozzy123nozzy** [comment]
@Pierre-Sassoulas I’m afraid that results appear to be the same. I'm not sure why thease are same results.
```
1) Increase limit-inference-result 100(default)->400
$ time env GDK_BACKEND=- pylint --limit-inference-results=400 -d C0114,W0611 test.py
                                                                                                                                        
--------------------------------------------------------------------                                                                     
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)                                                                     
                                                                                                                                         
                                                                                                                                         
real    0m14.096s                                                                                                                        
user    0m13.648s                                                                                                                        
sys     0m0.440s           

2) Decrease limit-inference-result 100(default)->10                                                                                                  
$ time env GDK_BACKEND=- pylint --limit-inference-results=10 -d C0114,W0611 test.py 
                                                                                                                                         
--------------------------------------------------------------------                                                                     
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)                                                                     
                                                                                                                                         
                                                                                                                                         
real    0m14.077s                                                                                                                        
user    0m13.709s                                                                                                                        
sys     0m0.360s                                                                                                                         
```

**Comment by mgrabovsky** [comment]
@Pierre-Sassoulas Thanks for your tips. I've managed to get rid of the “Unable to init server” message using `GDK_BACKEND=-`.

However, the problem with timeouts in GitHub workflows persists even with, for instance, `--limit-inference-results=5`. See [this recent run](https://github.com/abrt/gnome-abrt/runs/4936399015?check_suite_focus=true) for an example.

**Comment by Pierre-Sassoulas** [comment]
Thank you for the detail. It seems github action is killing the process after 180s, but do you manage to run in less than 180s locally ? It's entirely possible that pylint is taking more than 180s and would end normally in longer than that. Depending on the size of your codebase it could even be normal and not something we can easily optimize.

**Comment by mgrabovsky** [comment]
Pylint finishes in about 30 seconds on my machine. The runtime on GitHub seems to be a random variable. [Here's a run](https://github.com/abrt/gnome-abrt/runs/4936312746?check_suite_focus=true) that took 120 seconds and thus did not time out. (It's the same commit as before, just triggered by a PR instead of a push to master. I've seen this pattern multiple times before, though.)

**Comment by Pierre-Sassoulas** [comment]
I don't know how to check that are the specs of the machine for a github actions pipeline, but it would be interesting to see if there's any difference between runs.
